toolchain/check/convert.cpp

```diff
@@ -0,0 +1,729 @@
+    if (value_type_node.kind() == SemIR::NodeKind::TupleType) {
+      return ConvertTupleToTuple(context, value_type_node, target_type_node,
+                                 value_id, target);
+    }
+  }
+
+  // A struct {.f_1: T_1, .f_2: T_2, ..., .f_n: T_n} converts to
+  // {.f_p(1): U_p(1), .f_p(2): U_p(2), ..., .f_p(n): U_p(n)} if
+  // (p(1), ..., p(n)) is a permutation of (1, ..., n) and each Ti converts
+  // to Ui.
+  if (target_type_node.kind() == SemIR::NodeKind::StructType) {
+    auto value_type_node = context.semantics_ir().GetNode(
+        context.semantics_ir().GetTypeAllowBuiltinTypes(value_type_id));
+    if (value_type_node.kind() == SemIR::NodeKind::StructType) {
+      return ConvertStructToStruct(context, value_type_node, target_type_node,
+                                   value_id, target);
+    }
+  }
+
+  // A tuple (T1, T2, ..., Tn) converts to [T; n] if each Ti converts to T.
+  if (target_type_node.kind() == SemIR::NodeKind::ArrayType) {
+    auto value_type_node = context.semantics_ir().GetNode(
+        context.semantics_ir().GetTypeAllowBuiltinTypes(value_type_id));
+    if (value_type_node.kind() == SemIR::NodeKind::TupleType) {
+      return ConvertTupleToArray(context, value_type_node, target_type_node,
+                                 value_id, target);
+    }
+  }
+
+  if (target.type_id == SemIR::TypeId::TypeType) {
+    // A tuple of types converts to type `type`.
+    // TODO: This should apply even for non-literal tuples.
+    if (value.kind() == SemIR::NodeKind::TupleLiteral) {
+      auto tuple_block_id = value.GetAsTupleLiteral();
+      llvm::SmallVector<SemIR::TypeId> type_ids;
+      // If it is empty tuple type, we don't fetch anything.
+      if (tuple_block_id != SemIR::NodeBlockId::Empty) {
+        const auto& tuple_block =
+            context.semantics_ir().GetNodeBlock(tuple_block_id);
+        for (auto tuple_node_id : tuple_block) {
+          // TODO: This call recurses back into conversion. Switch to an
+          // iterative approach.
+          type_ids.push_back(
+              ExpressionAsType(context, parse_node, tuple_node_id));
+        }
+      }
+      auto tuple_type_id =
+          context.CanonicalizeTupleType(parse_node, std::move(type_ids));
+      return context.semantics_ir().GetTypeAllowBuiltinTypes(tuple_type_id);
+    }
+
+    // `{}` converts to `{} as type`.
+    // TODO: This conversion should also be performed for a non-literal value
+    // of type `{}`.
+    if (value.kind() == SemIR::NodeKind::StructLiteral &&
+        value.GetAsStructLiteral() == SemIR::NodeBlockId::Empty) {
+      value_id = context.semantics_ir().GetTypeAllowBuiltinTypes(value_type_id);
+    }
+  }
+
+  // No builtin conversion applies.
+  return value_id;
+}
+
+auto Convert(Context& context, Parse::Node parse_node, SemIR::NodeId expr_id,
+             ConversionTarget target) -> SemIR::NodeId {
+  auto orig_expr_id = expr_id;
```

Comment:
Consider:

```suggestion
             ConversionTarget target) -> SemIR::NodeId {
  auto semir = conetxt.semantic_ir();
  auto orig_expr_id = expr_id;
```

And using `semir` (or whatever name you prefer) below to make the code a bit shorter?

Reply:
Also happy to use this pattern for any of the functions where there are more than 3-ish calls.

Reply:
Done.